src/components/utils/test/timer_test.cc

```diff
@@ -43,214 +43,276 @@ namespace components {
+  EXPECT_EQ(kDefaultTimeout, timer.timeout());
+
+  // Wait for 3 calls
+  for (size_t i = 0; i < loops_count; ++i) {
+    shot_condition.Wait(shot_lock);
+    EXPECT_TRUE(timer.is_running());
   }
-  test_lock.Release();
-  test_timer.Stop();
 
-  EXPECT_FALSE(test_timer.is_running());
-  EXPECT_EQ(loops_count, task->GetCallsCount());
-}
+  timer.Stop();
+  EXPECT_FALSE(timer.is_running());
+  EXPECT_EQ(0u, timer.timeout());
 
-// {AKozoriz} : Disabled due correct realization of Timer
-// In case Start -> Immediately Stop | we have uncorrect behavior
-TEST_F(TimerTest, DISABLED_Start_Runned_RunnedWithNewTimeout) {
-  // Preconditions
-  timer::Timer test_timer(kTimerName, test_task_);
-  EXPECT_CALL(*test_task_, run()).Times(0);
-  // Actions
-  test_timer.Start(timeout_, single_shot_);
-  // Expects
-  ASSERT_EQ(timeout_, test_timer.timeout());
-  ASSERT_TRUE(test_timer.is_running());
-  // Actions
-  timeout_ = 1000u;
-  test_timer.Start(timeout_, single_shot_);
-  // Expects
-  ASSERT_EQ(timeout_, test_timer.timeout());
-  ASSERT_TRUE(test_timer.is_running());
-
-  test_timer.Stop();
-  ASSERT_FALSE(test_timer.is_running());
+  EXPECT_EQ(loops_count, task->calls_count());
 }
 
-TEST_F(TimerTest, DISABLED_Start_NotRunned_RunnedWithNewTimeout) {
-  // Preconditions
-  timer::Timer test_timer(kTimerName, test_task_);
-  // Expects
-  ASSERT_EQ(0u, test_timer.timeout());
-  ASSERT_FALSE(test_timer.is_running());
-  EXPECT_CALL(*test_task_, run()).Times(0);
-  // Actions
-  timeout_ = 1000u;
-  test_timer.Start(timeout_, single_shot_);
-  // Expects
-  ASSERT_EQ(timeout_, test_timer.timeout());
-  ASSERT_TRUE(test_timer.is_running());
-
-  test_timer.Stop();
-  ASSERT_FALSE(test_timer.is_running());
+// Restart
+
+TEST(TimerTest, Restart_NoLoop_NoCall) {
+  MockTimerTask* mock_task = new MockTimerTask();
+  EXPECT_CALL(*mock_task, run()).Times(0);
+
+  timer::Timer timer(kTimerName, mock_task);
+
+  timer.Start(kDefaultTimeout, kSingleShot);
+  EXPECT_TRUE(timer.is_running());
+  EXPECT_EQ(kDefaultTimeout, timer.timeout());
+
+  timer.Start(kDefaultTimeoutRestart, kSingleShot);
+  EXPECT_TRUE(timer.is_running());
+  EXPECT_EQ(kDefaultTimeoutRestart, timer.timeout());
 }
 
-TEST_F(TimerTest, DISABLED_Stop_FirstLoop_NoCall) {
-  // Preconditions
-  timer::Timer test_timer(kTimerName, test_task_);
-  // Expects
-  EXPECT_CALL(*test_task_, run()).Times(0);
-  // Actions
-  test_timer.Start(10000u, single_shot_);
-  test_timer.Stop();
+TEST(TimerTest, Restart_Loop_NoCall) {
+  MockTimerTask* mock_task = new MockTimerTask();
+  EXPECT_CALL(*mock_task, run()).Times(0);
+
+  timer::Timer timer(kTimerName, mock_task);
+
+  timer.Start(kDefaultTimeout, !kSingleShot);
+  EXPECT_TRUE(timer.is_running());
+  EXPECT_EQ(kDefaultTimeout, timer.timeout());
 
-  EXPECT_FALSE(test_timer.is_running());
+  timer.Start(kDefaultTimeoutRestart, !kSingleShot);
+  EXPECT_TRUE(timer.is_running());
+  EXPECT_EQ(kDefaultTimeoutRestart, timer.timeout());
 }
```

Comment:
@Kozoriz we don't need to call `Stop()` here. Timer is local object and it will be destroyed immediately.